zerotier/files/etc/init.d/zerotier

```diff
@@ -7,113 +7,94 @@ USE_PROCD=1
+	config_get secret 'global' 'secret'
+	config_get local_conf_path 'global' 'local_conf_path'
+	config_get config_path 'global' 'config_path'
+	config_get_bool copy_config_path 'global' 'copy_config_path' 0
+
+	if [ ${enabled} -eq 0 ]; then
 		echo "disabled in /etc/config/zerotier"
-		return 1
 	fi
 
-	config_get config_path $cfg 'config_path'
-	config_get port $cfg 'port'
-	config_get secret $cfg 'secret'
-	config_get local_conf $cfg 'local_conf'
-	config_get_bool copy_config_path $cfg 'copy_config_path' 0
-
-	path=${CONFIG_PATH}_$cfg
-
 	# Remove existing link or folder
-	rm -rf $path
+	rm -rf "${CONFIG_PATH}"
 
-	# Create link or copy files from CONFIG_PATH to config_path
-	if [ -n "$config_path" -a "$config_path" != "$path" ]; then
-		if [ ! -d "$config_path" ]; then
-			echo "ZeroTier config_path does not exist: $config_path" 1>&2
+	# Create link or copy files from config_path to CONFIG_PATH
+	if [ -n "${config_path}" -a "${config_path}" != "${path}" ]; then
+		if [ ! -d "${config_path}" ]; then
+			echo "ZeroTier config_path does not exist: ${config_path}" 1>&2
 			return
 		fi
 
-		# ensure that the target exists
-		mkdir -p $(dirname $path)
-
-		if [ "$copy_config_path" = "1" ]; then
-			cp -r $config_path $path
+		if [ ${copy_config_path} -eq 1 ]; then
+			cp -r "${config_path}" "${CONFIG_PATH}"
 		else
-			ln -s $config_path $path
+			ln -s "${config_path}" "${CONFIG_PATH}"
 		fi
 	fi
 
-	mkdir -p $path/networks.d
-
-	# link latest default config path to latest config path
-	rm -f $CONFIG_PATH
-	ln -s $path $CONFIG_PATH
-
-	if [ -n "$port" ]; then
-		args="$args -p${port}"
+	if [ -f "${local_conf_path}" ]; then
+		ln -s "${local_conf_path}" "${CONFIG_PATH}"/local.conf
 	fi
 
-	if [ -z "$secret" ]; then
-		echo "Generate secret - please wait..."
-		local sf="/tmp/zt.$cfg.secret"
-
-		zerotier-idtool generate "$sf" > /dev/null
-		[ $? -ne 0 ] && return 1
-
-		secret="$(cat $sf)"
-		rm "$sf"
+	if [ -n "${port}" ]; then
+		args="${args} -p${port}"
+	fi
 
-		uci set zerotier.$cfg.secret="$secret"
+	if [ -z "${secret}" ]; then
+		echo -n "Generating secret - please wait... "
+		secret="$(zerotier-idtool generate)"
+		[ ${?} -ne 0 ] && return 1
+		uci set zerotier.global.secret="${secret}"
 		uci commit zerotier
+		echo "done."
 	fi
 
-	if [ -n "$secret" ]; then
-		echo "$secret" > $path/identity.secret
+	if [ -n "${secret}" ]; then
+		echo "${secret}" > "${CONFIG_PATH}"/identity.secret
 		# make sure there is not previous identity.public
-		rm -f $path/identity.public
-	fi
-
-	if [ -f "$local_conf" ]; then
-		ln -s "$local_conf" $path/local.conf
+		rm -f "${CONFIG_PATH}"/identity.public
 	fi
 
-	add_join() {
-		# an (empty) config file will cause ZT to join a network
-		touch $path/networks.d/$1.conf
-	}
-
-	config_list_foreach $cfg 'join' add_join
+	mkdir -p "${CONFIG_PATH}"/networks.d
+	config_foreach join_network network
 
 	procd_open_instance
-	procd_set_param command $PROG $args $path
+	procd_set_param command ${PROG} ${args}
```

Comment:
Wouldn't `procd_set_param command "$PROG" "$args"` be safer here?

Reply:
We could put `${PROG}` in quotation marks, but we would not gain much since it is a variable defined by the script itself at the beginning (`PROG=/usr/bin/zerotier-one`) and nobody can overwrite it.

The same is true for `${args}`. It is a local variable of `start_service` that is defined in a safe way and that is already in quotes at the time of the definition [`args="${args} -p${port}"`](https://github.com/mwarning/zerotier-openwrt/pull/120/files#diff-500e5f9436fc26df6e76fbff4e9e3abf7b95b6de567fe91764888e3e2877e5b1R68). In fact you could even modify that `if` to simply be:
```sh
if [ -n "${port}" ]; then
    args="-p${port}"
fi
```
Since at the end either `${args}` has no value or it contains `-pXXX`. If you want I can modify that.

Reply:
fair enough